plutus-conformance/src/PlutusConformance/Common.hs

```diff
@@ -0,0 +1,134 @@
+        Just tests -> tests
+        Nothing -> error $ unlines
+            ["mkTestContents: Cannot run the tests because the number of input and output programs are not the same. "
+            , "Number of input files: "
+            , show (length lProgs)
+            , " Number of output files: "
+            , show (length lRes)
+            , " Make sure all your input programs have an accompanying .expected file."
+            ]
+
+{- | The default shown text when a parse error occurs.
+We don't want to show the detailed parse errors so that
+users of the test suite can produce this expected outputs more easily. -}
+shownParseError :: T.Text
+shownParseError = "parse error"
+
+-- | The default shown text when evaluation fails.
+shownEvaluationFailure :: T.Text
+shownEvaluationFailure = "evaluation failure"
+
+-- For UPLC evaluation tests
+
+type UplcProg = UPLC.Program Name DefaultUni DefaultFun ()
+
+termToProg :: UPLC.Term Name DefaultUni DefaultFun () -> UplcProg
+termToProg = UPLC.Program () (defaultVersion ())
+
+{- using the unsafe version of evaluate here so that it has a more generic signature.
+Any exceptions will be caught for any input runner in the tests, including our `evalUplcProg`.
+ -}
+evalUplcProg :: UplcProg -> EvaluationResult UplcProg
+evalUplcProg p =
+    fmap
+        termToProg
+        (unsafeEvaluateCekNoEmit defaultCekParameters (UPLC._progTerm p))
+
+{- | Run the inputs with the runner and return the results, in `Text`.
+When fail, the results are the default texts for parse error or evaluation failure. -}
+mkResult ::
+    -- | The `runner` to run the test inputs with.
+    (UplcProg -> EvaluationResult UplcProg) ->
+    -- | The result of parsing.
+    Either ParserErrorBundle (UPLC.Program Name DefaultUni DefaultFun SourcePos) ->
+    -- | The result in `Text`.
+        IO T.Text
+mkResult _ (Left (ParseErrorB _err)) = pure shownParseError
+mkResult runner (Right prog)        = do
+    maybeException <- try (evaluate $ runner (() <$ prog)):: IO (Either SomeException (EvaluationResult UplcProg))
+    case maybeException of
+        Left _                  -> pure shownEvaluationFailure
+        -- it doesn't matter how the evaluation fail, they're all "evaluation failure"
+        Right EvaluationFailure -> pure shownEvaluationFailure
+        Right a                 -> pure $ render $ pretty a
+
+-- | The default parser to parse the inputs.
+parseTxt ::
+    T.Text ->
+    Either ParserErrorBundle (UPLC.Program Name DefaultUni DefaultFun SourcePos)
+parseTxt resTxt = runQuoteT $ UPLC.parseProgram resTxt
+
+-- | Build the test tree given a list of `TestContent` and the runner.
+-- TODO maybe abstract this for other tests too if it takes in `mkResult` and `runner`.
+mkTestCases :: [TestContent] -> (UplcProg -> EvaluationResult UplcProg) -> IO TestTree
+mkTestCases lTest runner = do
+    results <- for lTest (mkResult runner . parseTxt . inputProg)
```

Comment:
Note that this _runs the test_ when _constructing the test tree_. This is not really what you're supposed to do, you're supposed to run the test when you... run the test! I'm not sure how much it matters in this case, but it'll make the progress of the test suite weird, and it may mess up the exception handling.

I think the solution is again in turning `TestContent` into a proper test specificaiton. That means it'll probably need to _include_ the test runner function, so that it can run it when the test is actually run, rather than when it's constructed.